x/stakeibc/keeper/icqcallbacks_pool_price.go

```diff
@@ -0,0 +1,95 @@
+package keeper
+
+import (
+	"fmt"
+
+	errorsmod "cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/gogoproto/proto"
+
+	"github.com/Stride-Labs/stride/v16/utils"
+	icqtypes "github.com/Stride-Labs/stride/v16/x/interchainquery/types"
+	"github.com/Stride-Labs/stride/v16/x/stakeibc/types"
+)
+
+// PoolPriceCallback is a callback handler for PoolPrice query.
+// The query response returns an Osmosis TwapRecord for the associated pool denom's
+//
+// The assets in the response are identified by indicies and are sorted alphabetically
+// (e.g. if the two denom's are ibc/AXXX, and ibc/BXXX,
+// then Asset0Denom is ibc/AXXX and Asset1Denom is ibc/BXXX)
+//
+// The price fields (P0LastSpotPrice and P1LastSpotPrice) represent the relative
+// ratios of tokens in the pool
+//
+//	P0LastSpotPrice gives the ratio of Asset0Denom / Asset1Denom
+//	P1LastSpotPrice gives the ratio of Asset1Denom / Asset0Denom
+//
+// When storing down the price, we want to denominate the price of the TargetDenom,
+// relative to the price of the RewardDenom, which means we have to take the inverse
+// from the response
+func PoolPriceCallback(k Keeper, ctx sdk.Context, args []byte, query icqtypes.Query) error {
+	k.Logger(ctx).Info(utils.LogICQCallbackWithHostZone(query.ChainId, ICQCallbackID_PoolPrice,
+		"Starting pool spot price callback, QueryId: %vs, QueryType: %s, Connection: %s", query.Id, query.QueryType, query.ConnectionId))
+
+	chainId := query.ChainId // should be the tradeZoneId, used in logging
+
+	// Unmarshal the query response args, should be a TwapRecord type
+	var twapRecord types.OsmosisTwapRecord
+	err := twapRecord.Unmarshal(args)
+	if err != nil {
+		return errorsmod.Wrap(err, "unable to unmarshal the query response")
+	}
+
+	// Unmarshal the callback data containing the tradeRoute we are on
+	var tradeRoute types.TradeRoute
+	if err := proto.Unmarshal(query.CallbackData, &tradeRoute); err != nil {
+		return errorsmod.Wrapf(err, "unable to unmarshal trade reward balance callback data")
+	}
+
+	// Confirm the denom's from the query response match the denom's in the route
+	if err := AssertTwapAssetsMatchTradeRoute(twapRecord, tradeRoute); err != nil {
+		return err
+	}
+
+	// Get the associate "SpotPrice" from the twap record, based on the asset ordering
+	// The "SpotPrice" is actually a ratio of the assets in the pool
+	var ratioOfHostToRewardDenom sdk.Dec
+	if twapRecord.Asset0Denom == tradeRoute.TargetDenomOnTradeZone {
+		ratioOfHostToRewardDenom = twapRecord.P0LastSpotPrice
+	} else {
+		ratioOfHostToRewardDenom = twapRecord.P1LastSpotPrice
+	}
+
+	// We want the price of the host denom in terms of the reward denom,
+	// so we can use the ratio above to determine the price, via (1 / RatioOfHostToReward)
+	hostPrice := sdk.OneDec().Quo(ratioOfHostToRewardDenom)
```

Comment:
Why involve this extra math which could make weird rounding issues (and should add an extra div by 0 check), from the definitions can't you just reverse the p0 and the p1 spot prices in the if statement above to already get what you want?

Reply:
yeah my thinking was that this would be a bit more readable (since seeing "if asset0, set price1" would be a bit confusing

But rounding issues and the div by 0 check are good reasons to not do the division. I'll update this!